aten/src/ATen/native/mobile/cpu/internal/Clamp.cpp

```diff
@@ -0,0 +1,208 @@
+#include <ATen/native/mobile/cpu/Engine.h>
+
+#ifdef USE_XNNPACK
+
+#include <ATen/native/mobile/cpu/internal/Allocator.h>
+#include <ATen/native/mobile/cpu/internal/Common.h>
+
+namespace at {
+namespace native {
+namespace mobile {
+namespace cpu {
+namespace internal {
+namespace clamp {
+
+struct Context final {
+  Operator clamp_op;
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32
+//  - relu
+//  - relu6
+//  - leaky relu
+//  - hard tanh
+//  - or any other operator that can be expressed as a clamp
+
+bool available(
+    const int64_t input_channels,
+    const int64_t output_channels,
+    const Scalar output_min,
+    const Scalar output_max) {
+         // Mobile
+  return mobile::cpu::available() &&
+         // Input
+         (input_channels > 0) &&
+         // Output
+         (output_channels > 0) &&
+         // Input == Output
+         (input_channels == output_channels) &&
+         // Output Min / Max
+         (output_min.isIntegral(true) || output_min.isFloatingPoint()) &&
+         (output_max.isIntegral(true) || output_max.isFloatingPoint()) &&
+         (output_max.to<float>() > output_min.to<float>());
+}
+
+Context create(
+    const int64_t input_channels,
+    const int64_t output_channels,
+    const Scalar output_min,
+    const Scalar output_max) {
+  TORCH_INTERNAL_ASSERT(
+      available(input_channels, output_channels, output_min, output_max),
+      "mobile::cpu::clamp not available!");
+
+  xnn_operator_t clamp_op{};
+
+  const xnn_status create_status = xnn_create_clamp_nc_f32(
+      input_channels,         // channels,
+      input_channels,         // input_pixel_stride
+      output_channels,        // output_pixel_stride
+      output_min.to<float>(), // output_min
+      output_max.to<float>(), // output_max
+      0u,                     // flags
+      &clamp_op);             // operator
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == create_status,
+      "xnn_create_clamp_nc_f32 failed!");
+
+  return Context{
+    Operator(clamp_op),
+  };
+}
+
+bool usable(const Tensor& input) {
+         // Input
+  return (0 < input.ndimension()) &&
```

Comment:
It is also pretty weird that XNNPACK doesn't handle scalar inputs properly

Reply:
@ezyang what do you mean? Scalar inputs should work just fine